[stdlib] Modernize Set and Dictionary internals

lorentey opened the pull request:
Clean up Set and Dictionary internals to improve readability and to prepare for upcoming resiliency improvements.

- Nest public iterator types under the corresponding collection type, with a compatibility typealias.
    `SetIterator<Element>` ⟹ `Set<Element>.Iterator`
    `DictionaryIterator<Key, Value>` ⟹ `Dictionary<Key, Value>.Iterator`

- Nest auxiliary types under the corresponding higher-level type. Remove confusing `Buffer` suffix. Unify naming of enums representing native/cocoa variants.
    `_VariantSetBuffer<Element>` ⟹ `Set<Element>._Variant`
    `SetIndexRepresentation<Element>` ⟹ `Set<Element>.Index._Variant`
    `SetIteratorRepresentation<Element>` ⟹ `Set<Element>.Iterator._Variant`

    `_NativeSetBuffer<Element>` ⟹ `_NativeSet<Element>`
    `_NativeSetIndex<Element>` ⟹ `_NativeSet<Element>.Index`

    `_CocoaSetBuffer` ⟹ `_CocoaSet`
    `_CocoaSetIndex` ⟹ `_CocoaSet.Index`
    `_CocoaSetIterator` ⟹ `_CocoaSet.Iterator`

    (Ditto for `Dictionary`.)

- Add explicit types representing native iterators: `_NativeSet<Element>.Iterator` and `_NativeDictionary<Key, Value>.Iterator`. Previously these were exploded into associated values on the native case of the top-level iterator variant.

- Make Cocoa helper class names a little bit more helpful in Cocoa context.
    `_NativeSetNSEnumerator` ⟹ `_SwiftSetNSEnumerator`
    `_NativeDictionaryNSEnumerator` ⟹ `_SwiftDictionaryNSEnumerator`

- Adjust internal property and variable names accordingly.
- To enhance readability, remove internal typealiases wherever possible.
- Audit inlinability.

rdar://problem/42585384

lorentey committed:
[stdlib] Set, Dictionary: Review variant enums

- Rename variant enum types and move them into the corresponding struct:
  _VariantSetBuffer => Set._Variant
  _VariantDictionaryBuffer => Dictionary._Variant

  I found the buffer suffix did not positively contribute to understanding these types. And the extra indirection of the typealiases wasn't exactly helpful, either.

- Remove internal typealiases that resolved to these.

lorentey committed:
[stdlib] Set, Dictionary: Remove obsolete operator definitions.

lorentey committed:
[stdlib][NFC] Shorten notes about SR-6437

lorentey committed:
[stdlib] Set, Dictionary: Review native & cocoa representations

- Remove buffer suffix from type, property & variable names
  struct _NativeSetBuffer => _NativeSet
  struct _NativeDictionaryBuffer => _NativeDictionary
  struct _CocoaSetBuffer => _CocoaSet
  struct _CocoaDictionaryBuffer => _CocoaDictionary

- Remove internal typealiases related to these types
- Move nativeDelete, nativeMapValues from variant enum to corresponding struct

lorentey committed:
[stdlib] Set, Dictionary: Review enumerators

_NativeSetNSEnumerator => _SwiftSetNSEnumerator
_NativeDictionaryNSEnumerator => _SwiftDictionaryNSEnumerator

lorentey committed:
[stdlib] Set, Dictionary: Review indices

- Nest native/cocoa index types under their corresponding struct
- Move *IndexRepresentation enums under the corresponding Index type, renamed to _Variant.
- Rename Index._value to Index._variant.
- Remove internal typealiases.
- Remove underscore prefix from enum case names.
- _nativeIndex => _asNative, _cocoaIndex => _asCocoa
- Audit inlinability.

lorentey committed:
[stdlib] Set, Dictionary: Audit inlinability of storage classes

- Remove @inlinable from @objc methods.
- _UnmanagedAnyObjectArray: make internal.

lorentey committed:
[stdlib] Set, Dictionary: Review iterators

- DictionaryIterator => Dictionary.Iterator
- SetIterator => Set.Iterator

- Introduce an explicit struct for native iterators
- _CocoaDictionaryIterator => _CocoaDictionary.Iterator
- _CocoaSetIterator => _CocoaSet.Iterator
- Move representation enums under {Set,Dictionary}.Iterator as _Variant.
- Inlinability audit

airspeedswift commented on stdlib/public/core/Set.swift:
```diff
@@ -390,7 +382,7 @@ extension Set: Collection {
   /// If the set is empty, the value of this property is `nil`.
-  @inlinable // FIXME(sil-serialize-all)
+  @inlinable
   public var first: Element? {
```
I wonder if this really does materially perform better than the default.

lorentey replied:
The default creates an iterator, which complicates the native case, but improves the bridged case -- `_CocoaSet.startIndex` is currently O(n), while the iterator uses fast enumeration. 😦

Ideally this should dispatch to `_variant` and do the right thing in each case. Or we could just remove it.

lorentey replied:
Actually, it turns out this is the correct way to implement this, despite the performance hit. 

Bridged indices are *currently* based on `-[NSDictionary allKeys]`, which is not guaranteed to use the same element ordering as an enumerator. (Which indicates Dictionary/Set iterators may sometimes return elements in a different order than indices. 😭)

I think we should change indices to use enumerators, and soon.

lorentey committed:
[stdlib] Set, Dictionary: Review builders

- Inlinability audit
- Create a _NativeSet/_NativeDictionary directly, without extracting it from a Set/Dictionary
- Swap the final result with an empty collection.

lorentey committed:
[stdlib] Set, Dictionary: Audit one more round of inlinability attributes

lorentey committed:
[stdlib] Set, Dictionary: Move compatibility typealiases to end of file

lorentey committed:
[stdlib] Set, Dictionary: Remove remaining internal typealiases

lorentey committed:
[stdlib] Dictionary: Add implementation for first

This is primarily for documentation purposes, although the default implementation (based on an iterator) may not return the correct value for bridged dictionaries with exotic implementations.

lorentey commented:
@swift-ci please test

lorentey commented:
The benchmarks shouldn't be affected, although there were a couple changes to inlinability.
@swift-ci please benchmark

lorentey committed:
[test] Update Mirror tests for new Set/Dictionary internals

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please test

lorentey committed:
[stdlib] _UnsafeBitMap: make internal; force-inline members

A Dictionary.removeValue(forKey:) benchmark regressed 35% because recent changes in this PR caused an _UnsafeBitMap member to not be inlined in its implementation. (This was probably triggered by moving a method from Dictionary._Variant to _NativeDictionary.)
Add @inline(__always) to _UnsafeBitMap members.

While we’re at it, make _UnsafeBitMap @usableFromInline. It’s only public for testing purposes.

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci test

lorentey commented:
@swift-ci benchmark

lorentey commented:
The `DictionaryRemove` benchmark regression looks legit: the compiler decided not to inline some `_UnsafeBitmap` members. Bitmap operations should always be inlined; I expect that enforcing this will fix the regression.

lorentey commented:
The plot thickens!

lorentey commented:
Please test with following PR:
https://github.com/apple/swift-lldb/pull/797

@swift-ci please test

lorentey committed:
[stdlib] Set, Dictionary: Prefer collection-agnostic names for stored property names

This makes it easier to define lldb data formatters for these types.

lorentey commented:
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/813

@swift-ci test

lorentey committed:
[test] Update api-digester source compatibility tests

DictionaryIterator & SetIterator have been renamed Dictionary.Iterator and Set.Iterator, with compatibility typealiases.

lorentey commented:
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/813

@swift-ci test

lorentey commented:
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/813

@swift-ci test

lorentey commented:
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/813

@swift-ci test

lorentey commented:
Please test with the following PR:
https://github.com/apple/swift-lldb/pull/813

@swift-ci test

lorentey committed:
[stdlib] Set, Dictionary: Simplify _find(_:,startBucket:)

The startBucket argument is _bucket(key) in the vast majority of cases. Add an overload with this default value.

lorentey committed:
[stdlib] Set, Dictionary: Make internal representations conform to Sequence

lorentey committed:
[stdlib] Set, Dictionary: Review _SwiftDeferredNS*

lorentey committed:
[stdlib] Implement Set.contains without retrieving the stored value

lorentey committed:
[stdlib] Make Set.isSubset(of:) self-contained; remove standalone function for same

lorentey commented:
@swift-ci please test and merge

lorentey commented:
@swift-ci please test and merge

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please test

airspeedswift commented:
@swift-ci please smoke test compiler performance

lorentey commented:
@swift-ci please smoke test compiler performance

airspeedswift commented:
That first wall time regression was just noise, then, not a problem.
